libraries/AC_Avoidance/AP_OABendyRuler.cpp

```diff
@@ -232,6 +244,159 @@ bool AP_OABendyRuler::calc_margin_from_polygon_fence(const Location &start, cons
     return true;
 }
 
+// calculate minimum distance between a path and all exclusion polygons
+// on success returns true and updates margin
+bool AP_OABendyRuler::calc_margin_from_exclusion_polygon(const Location &start, const Location &end, float &margin)
+{
+    // exit immediately if polygon fence is not enabled
+    const AC_Fence *fence = AC_Fence::get_singleton();
+    if (fence == nullptr) {
+        return false;
+    }
+
+    // exclusion polygons enabled along with polygon fences
+    if ((fence->get_enabled_fences() & AC_FENCE_TYPE_POLYGON) == 0) {
+        return false;
+    }
+
+    // return immediately if no exclusion polygons
+    const uint16_t num_exclusion_polygons = fence->polyfence().get_exclusion_polygon_count();
+    if (num_exclusion_polygons == 0) {
+        return false;
+    }
+
+    // convert start and end to offsets from EKF origin
+    Vector2f start_NE, end_NE;
+    if (!start.get_vector_xy_from_origin_NE(start_NE) || !end.get_vector_xy_from_origin_NE(end_NE)) {
+        return false;
+    }
+
+    // iterate through exclusion polygons and calculate minimum margin
+    bool margin_updated = false;
+    for (uint16_t i = 0; i < num_exclusion_polygons; i++) {
+        uint16_t num_points;
+        const Vector2f* boundary = fence->polyfence().get_exclusion_polygon(i, num_points);
+        if (num_points < 3) {
+            // ignore exclusion polygons with less than 3 points
+            continue;
+        }
+
+        // if start is inside the polygon the margin's sign is reversed
+        const float sign = Polygon_outside(start_NE, boundary, num_points) ? 1.0f : -1.0f;
+
+        // calculate min distance (in meters) from line to polygon
+        float margin_new = sign * Polygon_closest_distance_line(boundary, num_points, start_NE, end_NE) * 0.01f;
+        if (!margin_updated || (margin_new < margin)) {
+            margin_updated = true;
+            margin = margin_new;
+        }
+    }
+
+    return margin_updated;
+}
+
+// calculate minimum distance between a path and all exclusion circles
+// on success returns true and updates margin
+bool AP_OABendyRuler::calc_margin_from_exclusion_circle(const Location &start, const Location &end, float &margin)
+{
+    // exit immediately if fence is not enabled
+    const AC_Fence *fence = AC_Fence::get_singleton();
+    if (fence == nullptr) {
+        return false;
+    }
+
+    // exclusion circles enabled along with polygon fences
+    if ((fence->get_enabled_fences() & AC_FENCE_TYPE_POLYGON) == 0) {
+        return false;
+    }
+
+    // return immediately if no exclusion circles
+    const uint8_t num_exclusion_circles = fence->polyfence().get_exclusion_circle_count();
+    if (num_exclusion_circles == 0) {
+        return false;
```

Comment:
maybe it's symantics but it "failed" to calculate the margin to exclusion circles because there are no exclusion circles.  basically it just tells the caller that it shouldn't try and use the margin returned.